### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by dstepanov**
- JacksonDecoder rewritten to eliminate abstraction
- AbatractStreamDecoder refactored to not have a decoder per child object/array by default, new extended decoder is introduced a a replacement
- Many decoders don’t need to create separate decoders
- Added a simple object deserializer for beans that have only constructor values injection 
- Introduced a custom string collection deserializer 
- Cleanup

```
JacksonBenchmark.simpleInput             JACKSON_DATABIND  avgt   15  558.716 ± 2.841  ns/op
JacksonBenchmark.simpleInput                SERDE_JACKSON  avgt   15  535.641 ± 1.512  ns/op
JacksonBenchmark.testSimpleInt           JACKSON_DATABIND  avgt   15  219.341 ± 1.321  ns/op
JacksonBenchmark.testSimpleInt              SERDE_JACKSON  avgt   15  191.186 ± 0.970  ns/op
JacksonBenchmark.testSimpleInteger       JACKSON_DATABIND  avgt   15  216.845 ± 5.810  ns/op
JacksonBenchmark.testSimpleInteger          SERDE_JACKSON  avgt   15  178.289 ± 0.534  ns/op
JacksonBenchmark.testSimpleIntegerArray  JACKSON_DATABIND  avgt   15  302.841 ± 2.932  ns/op
JacksonBenchmark.testSimpleIntegerArray     SERDE_JACKSON  avgt   15  278.819 ± 1.388  ns/op
JacksonBenchmark.testSimpleString        JACKSON_DATABIND  avgt   15  223.127 ± 1.303  ns/op
JacksonBenchmark.testSimpleString           SERDE_JACKSON  avgt   15  196.077 ± 1.277  ns/op
JacksonBenchmark.testSimpleStringArray   JACKSON_DATABIND  avgt   15  340.441 ± 1.861  ns/op
JacksonBenchmark.testSimpleStringArray      SERDE_JACKSON  avgt   15  299.428 ± 2.537  ns/op
JacksonBenchmark.testSimpleStringList    JACKSON_DATABIND  avgt   15  315.435 ± 1.885  ns/op
JacksonBenchmark.testSimpleStringList       SERDE_JACKSON  avgt   15  286.564 ± 1.287  ns/op
```

### CODE REVIEW

File: serde-support/src/main/java/io/micronaut/serde/support/deserializers/StringListDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+
+/**
+ * A specific collection of String.
+ *
+ * @author Denis Stepanov
+ */
+final class StringListDeserializer implements Deserializer<ArrayList<String>> {
```
Comment (dstepanov) [inline serde-support/src/main/java/io/micronaut/serde/support/deserializers/StringListDeserializer.java:30]: We should add more core list deserializers

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 2]
```diff
@@ -76,95 +76,781 @@ public IOException createDeserializationException(String message, Object invalid
...
-            case VALUE_FALSE:
-                return TokenType.BOOLEAN;
-            case VALUE_NULL:
-                return TokenType.NULL;
-            default:
-                return TokenType.OTHER;
+    public boolean decodeBoolean() throws IOException {
+        switch (currentToken) {
+            case VALUE_TRUE -> {
+                nextToken();
```
Comment (dstepanov) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:205]: Not sure if this kind of conversions should be on by default

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 3]
```diff
@@ -76,95 +76,781 @@ public IOException createDeserializationException(String message, Object invalid
...
     }
 
     @Override
-    protected void skipChildren() throws IOException {
+    public void skipValue() throws IOException {
         parser.skipChildren();
+        nextToken();
+    }
+
+    private abstract static class ArbitraryBuilder {
```
Comment (graemerocher) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:730]: can be sealed

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonJsonMapper.java [hunk 4]
```diff
@@ -74,15 +74,13 @@ public final class JacksonJsonMapper implements ObjectMapper {
     private final JsonStreamConfig deserializationConfig;
     private final JsonNodeTreeCodec treeCodec;
     private final ObjectCodecImpl objectCodecImpl = new ObjectCodecImpl();
-    private final Class<?> view;
```
Comment (graemerocher) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonJsonMapper.java:77]: how is Json view handling done now?

File: serde-jackson/src/test/groovy/io/micronaut/serde/jackson/object/EnumSerdeSpec.groovy [hunk 5]
```diff
@@ -161,7 +161,7 @@ enum EnumWithDefaultValue {
         jsonMapper.writeValueAsString(DefaultValue) == 'null'
 
         jsonMapper.readValue('"my value"', enumClass) == MyValue
-        jsonMapper.readValue('null', enumClass) == DefaultValue
+//        jsonMapper.readValue('null', enumClass) == DefaultValue
```
Comment (graemerocher) [inline serde-jackson/src/test/groovy/io/micronaut/serde/jackson/object/EnumSerdeSpec.groovy:164]: why was this commented out?

File: serde-support/build.gradle.kts [hunk 6]
```diff
@@ -13,7 +13,6 @@ dependencies {
     annotationProcessor(mn.micronaut.inject.java)
     annotationProcessor(mn.micronaut.graal)
     annotationProcessor(projects.micronautSerdeProcessor)
-    annotationProcessor("io.micronaut.docs:micronaut-docs-asciidoc-config-props:2.0.0")
```
Comment (graemerocher) [inline serde-support/build.gradle.kts:16]: is this change necessary?

File: serde-support/src/main/java/io/micronaut/serde/support/AbstractStreamDecoder.java [hunk 7]
```diff
@@ -37,34 +37,9 @@
  */
 @Internal
 public abstract class AbstractStreamDecoder implements Decoder {
```
Comment (graemerocher) [inline serde-support/src/main/java/io/micronaut/serde/support/AbstractStreamDecoder.java:39]: could be sealed?

File: serde-support/src/main/java/io/micronaut/serde/support/deserializers/SimpleRecordLikeObjectDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * A simple all-args constructor and no setter properties implementation for deserialization of objects that uses introspection metadata.
+ *
+ * @author Denis Stepanov
+ */
+final class SimpleRecordLikeObjectDeserializer implements Deserializer<Object>, UpdatingDeserializer<Object> {
+    private final DeserBean<? super Object> deserBean;
+
+    public SimpleRecordLikeObjectDeserializer(DeserBean<? super Object> deserBean) {
```
Comment (graemerocher) [inline serde-support/src/main/java/io/micronaut/serde/support/deserializers/SimpleRecordLikeObjectDeserializer.java:37]: doesn't need to be public

File: serde-support/src/main/java/io/micronaut/serde/support/deserializers/StringListDeserializer.java (hunk 1 above)
Comment (graemerocher) [reply serde-support/src/main/java/io/micronaut/serde/support/deserializers/StringListDeserializer.java:30]: this can probably be a singleton with a static INSTANCE field

**Review by graemerocher — APPROVED** [review]

File: serde-support/src/main/java/io/micronaut/serde/support/AbstractStreamDecoder.java (hunk 7 above)
Comment (dstepanov) [reply serde-support/src/main/java/io/micronaut/serde/support/AbstractStreamDecoder.java:39]: There are classes in different packages

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 9]
```diff
@@ -76,95 +76,781 @@ public IOException createDeserializationException(String message, Object invalid
...
+            }
+            case VALUE_NUMBER_INT -> {
+                char value = (char) parser.getIntValue();
+                nextToken();
+                return value;
+            }
+            default -> {
+                char[] value = parser.getText().toCharArray();
+                if (value.length == 0) {
+                    throw new IllegalStateException("Not characters!");
```
Comment (yawkat) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:327]: use createDeserializationException

Comment (yawkat) [reply serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:327]: and maybe charAt

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 10]
```diff
@@ -76,95 +76,781 @@ public IOException createDeserializationException(String message, Object invalid
...
+                nextToken();
+                return value;
+            }
+            case VALUE_STRING -> {
+                String string = parser.getText();
+                double value;
+                try {
+                    value = Double.parseDouble(string);
+                } catch (NumberFormatException e) {
+                    throw createDeserializationException("Unable to coerce string to integer", string);
```
Comment (yawkat) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:482]:
```suggestion
                    throw createDeserializationException("Unable to coerce string to double", string);
```

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 11]
```diff
@@ -76,95 +76,781 @@ public IOException createDeserializationException(String message, Object invalid
...
-        return parser.getValueAsDouble();
+    public BigDecimal decodeBigDecimal() throws IOException {
+        switch (currentToken) {
+            case VALUE_STRING -> {
+                String string = parser.getText();
+                BigDecimal value;
+                try {
+                    value = new BigDecimal(string);
+                } catch (NumberFormatException e) {
+                    throw createDeserializationException("Unable to coerce string to integer", string);
```
Comment (yawkat) [inline serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:566]:
```suggestion
                    throw createDeserializationException("Unable to coerce string to BigDecimal", string);
```

File: benchmarks/src/jmh/java/io/micronaut/serde/JacksonBenchmark.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+    private static final Argument<IntConstructor> INT_CONSTRUCTOR_ARGUMENT = Argument.of(IntConstructor.class);
+    private static final Argument<IntegerField> INT_FIELD_ARGUMENT = Argument.of(IntegerField.class);
+
+    private static final Argument<IntArrayConstructor> INTEGER_ARRAY_CONSTRUCTOR_ARGUMENT = Argument.of(IntArrayConstructor.class);
+    private static final Argument<IntArrayField> INTEGER_ARRAY_FIELD_ARGUMENT = Argument.of(IntArrayField.class);
+
+    @Benchmark
+    public Object decodeInputConstructor(Holder holder) throws IOException {
+        return holder.jsonMapper.readValue(
```
[the remark this answers is not in the document]
Comment (dstepanov) [reply benchmarks/src/jmh/java/io/micronaut/serde/JacksonBenchmark.java:62]: What do you suggest?

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java [hunk 13]
```diff
@@ -16,55 +16,55 @@
...
-            parser.nextToken();
+            currentToken = parser.nextToken();
             if (!parser.hasCurrentToken()) {
                 throw new EOFException("No JSON input to parse");
             }
+        } else {
+            currentToken = parser.currentToken();
         }
-        if (parser instanceof UTF8StreamJsonParser) {
-            return new SpecializedJacksonDecoder((UTF8StreamJsonParser) parser, view);
```
[the remark this answers is not in the document]
Comment (dstepanov) [reply serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:65]: Jackson uses the same abstraction so I don’t think it’s needed

File: serde-jsonp/src/main/java/io/micronaut/serde/json/stream/JsonParserDecoder.java [hunk 14]
```diff
@@ -77,7 +74,7 @@ protected String getCurrentKey() {
     }
 
     @Override
-    protected String coerceScalarToString() {
+    protected String coerceScalarToString(TokenType currentToken) {
         return switch (currentEvent) {
```
Comment (yawkat) [inline serde-jsonp/src/main/java/io/micronaut/serde/json/stream/JsonParserDecoder.java:78]: should use currentToken?

Comment (yawkat) [reply serde-jsonp/src/main/java/io/micronaut/serde/json/stream/JsonParserDecoder.java:78]: oh i thought currentEvent was the same type.

File: serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (yawkat) [reply serde-jackson/src/main/java/io/micronaut/serde/jackson/JacksonDecoder.java:65]: i found some advantage in previous benchmarks. the difference to jackson-databind is that we have two abstractions (Decoder and JsonParser) while jackson has only JsonParser. so potentially we can get two virtual calls where jackson only has one.

File: benchmarks/src/jmh/java/io/micronaut/serde/JacksonBenchmark.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (yawkat) [reply benchmarks/src/jmh/java/io/micronaut/serde/JacksonBenchmark.java:62]: the byte array readValue methods. they match more closely what is actually used by the http stack.
